src/ints/xms.cpp

```diff
@@ -102,392 +110,752 @@ struct XMS_MemMove{
 #pragma pack ()
 #endif
 
+// ***************************************************************************
+// Variables
+// ***************************************************************************
+
+static struct {
+	bool enable_global = false;
+
+	uint32_t enable_local_count = 0;
+} a20;
+
+constexpr uint32_t A20MaxLocalCount = UINT32_MAX;
+
+static struct {
+	// TODO: HMA support for applications is not yet available in the core
+
+	bool is_available = false;
+
+	bool dos_has_control = true;
+	bool app_has_control = false;
+
+	uint16_t min_alloc_size = 0;
+} hma;
+
+static struct {
+	bool is_available = false;
+} umb;
+
+static struct {
+	bool is_available = false;
+
+	RealPt callback = 0;
 
-Bitu XMS_EnableA20(bool enable) {
-	uint8_t val = IO_Read	(0x92);
-	if (enable) IO_Write(0x92,val | 2);
-	else		IO_Write(0x92,val & ~2);
-	return 0;
+	XMS_Block handles[NumXmsHandles] = {};
+} xms;
+
+// ***************************************************************************
+// Generic helper routines
+// ***************************************************************************
+
+static uint32_t get_num_pages(const uint32_t size_kb)
+{
+	// Number of pages needed for the given memory size in KB
+	return size_kb / 4 + ((size_kb & 3) ? 1 : 0);
+}
+
+static uint32_t get_mem_free_total_kb()
+{
+	return static_cast<uint32_t>(MEM_FreeTotal() * KilobytesPerPage);
+}
+
+static uint32_t get_mem_free_largest_kb()
+{
+	return static_cast<uint32_t>(MEM_FreeLargest() * KilobytesPerPage);
 }
 
-Bitu XMS_GetEnabledA20(void) {
-	return (IO_Read(0x92)&2)>0;
+static uint32_t get_mem_highest_address()
+{
+	return static_cast<uint32_t>((MEM_TotalPages() * MemPageSize) - 1);
+}
+
+static void warn_umb_realloc()
+{
+	static bool first_time = true;
+	if (first_time) {
+		first_time = false;
+		LOG_WARNING("XMS: UMB realloc not implemented");
+	}
+}
+
+// ***************************************************************************
+// Gate A20 support
+// ***************************************************************************
+
+static void a20_enable(const bool enable)
+{
+	uint8_t val = IO_Read(port_num_fast_a20);
+	if (enable) {
+		bit::set(val, bit::literals::b1);
+	} else {
+		bit::clear(val, bit::literals::b1);
+	}
```

Comment:
`set_to(...)` could be used here.

From:
```c++
if (enable) {
	bit::set(val, bit::literals::b1);
} else {
	bit::clear(val, bit::literals::b1);
}
```
To:
```c++
// set val's bit-1 to 'enable' (on/off)
bit::set_to(val, bit::literals::b1, enable);
```

Reply:
(just realized there were no unit tests for `"set_to(..)`, so I've [added them now](https://github.com/dosbox-staging/dosbox-staging/actions/runs/4645959685/jobs/8222092631) and are passing)

Reply:
Ok, changed.